vechain integration

davidecarpini opened the pull request:
### 📝 Description
The changes enable native vechain support for the Ledger Live application. Ledger Live users should be able to manage their vechain accounts, send/receive VET and VTHO and view their transaction history, crypto and fiat balances.

### ❓ Context
Impacted projects: ledgerjs, ledger-live-common, ledger-live-desktop, ledger-live-mobile

### ✅ Checklist
 Test coverage
 Atomic delivery
 No breaking changes

### 🚀 Expectations to reach
Add VET currency
Add VTHO token, listed as ERC20 token
Syncing VET accounts and VTHO subaccounts with the blockchain displaying the correct balances and operations list (through cli/lld/llm)
Send VET or VTHO transaction (through cli/lld/llm)
Receive VET or VTHO
Hide the graph for VTHO
Correctly displaying all the transaction information with correct links to the explorer

### 📸 Demo

![235204731-974b7b78-2795-4381-9ea1-d3b43a1f9ff4](https://github.com/LedgerHQ/ledger-live/assets/5785838/8adfcb3b-40c7-4f64-8b27-7fa0fcf88477)
![235204441-a81fd0e2-c4d6-4caf-b005-00bc314a836d](https://github.com/LedgerHQ/ledger-live/assets/5785838/048cc145-bbce-4afc-add3-5283956eb2bf)
![235204180-3b767fec-f566-4480-bc82-c862ac2659f9](https://github.com/LedgerHQ/ledger-live/assets/5785838/f04898e6-9db2-44ff-9cf4-838804658555)
![235204099-7f668c21-3b11-48e2-8ee3-f38726770ca1](https://github.com/LedgerHQ/ledger-live/assets/5785838/f7d0c0a9-d333-4a15-96d3-b12522a8e03e)
![235204083-1bedfc7e-08bd-415c-b3d2-c425e20d1d9b](https://github.com/LedgerHQ/ledger-live/assets/5785838/0073dbd8-f94f-4abc-b86c-fb3725ec4b03)

https://github.com/LedgerHQ/ledger-live/assets/5785838/1d9013d0-febf-4fb0-8fc6-2b349d0dab6d


https://github.com/LedgerHQ/ledger-live/assets/5785838/e4dfaa1b-e0a6-4fee-9d71-fcce2b8bbf60


https://github.com/LedgerHQ/ledger-live/assets/5785838/fc3a7bb8-6f39-4608-a754-0feaa1eb6812

<img width="1279" alt="235185649-c21259df-f753-40e8-8717-5aedfbc85fbf" src="https://github.com/LedgerHQ/ledger-live/assets/5785838/27f2660b-197e-49e6-81a1-4d28abeff1c7">
<img width="1280" alt="235185630-3d1901b3-1d4e-4146-b392-55c87dc04916" src="https://github.com/LedgerHQ/ledger-live/assets/5785838/41217af7-071e-452c-ada4-e83aa92c4ed1">
<img width="1013" alt="235185655-afd33036-b121-4a3f-827c-c0d14a146c1f" src="https://github.com/LedgerHQ/ledger-live/assets/5785838/0ac13326-36ed-442b-b6f4-6f0638c08ec9">

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/js-synchronisation.ts:
```diff
@@ -0,0 +1,91 @@
+
+const getAccountShape: GetAccountShape = async info => {
+  const { initialAccount, currency, derivationMode } = info;
+  let { address } = info;
+  address = eip55.encode(address);
```
```suggestion
  const address = eip55.encode(info.address);
```

davidecarpini replied:
thanks changed

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/api/sdk.ts:
```diff
@@ -0,0 +1,144 @@
+  return data;
+};
+
+export const getLastBlock = async (): Promise<{ number: number }> => {
```
```suggestion
export const getLastBlockHeight = async (): Promise<number> => {
```

davidecarpini replied:
changed thanks

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/js-synchronisation.ts:
```diff
@@ -0,0 +1,91 @@
+  const newOperations = await getOperations(accountId, address, startAt);
+
+  //Get last token operations
+  const vthoAccountId = accountId + "+" + encodeURIComponent("vechain/vtho");
```
Please use `encodeTokenAccountId` here

davidecarpini replied:
added thanks

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/js-synchronisation.ts:
```diff
@@ -0,0 +1,91 @@
+
+  //Get last token operations
+  const vthoAccountId = accountId + "+" + encodeURIComponent("vechain/vtho");
+  const VTHOoperations = await getTokenOperations(
```
(nitpicking on the case)
```suggestion
  const vthoOperations = await getTokenOperations(
```

davidecarpini replied:
changed thanks

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/js-synchronisation.ts:
```diff
@@ -0,0 +1,91 @@
+    creationDate: minDate != -1 ? new Date(minDate) : new Date(),
+    spendableBalance: new BigNumber(balance),
+    operationsCount: operations.length,
+    operations: operations,
```
```suggestion
    operations,
```

davidecarpini replied:
changed thanks

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/js-synchronisation.ts:
```diff
@@ -0,0 +1,91 @@
+    ],
+  };
+
+  return { ...shape, operations };
```
Unneeded, you can just `return shape;`

davidecarpini replied:
thanks changed

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/js-transaction.ts:
```diff
@@ -0,0 +1,135 @@
+    family: "vechain",
+    body: {
+      chainTag,
+      // placeholder, if "empty" returns error on send modal open
```
I don't understand this comment

davidecarpini replied:
Mirko wrote it, seems like that blockref has to be initialized because otherwise, it will throw an error

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/js-transaction.ts:
```diff
@@ -0,0 +1,135 @@
+    }
+  }
+
+  const gas = await estimateGas({
+    ...transaction,
+    body: { ...transaction.body, clauses: clauses },
+  });
+
+  const estimatedFees = await calculateFee(new BigNumber(gas), transaction.body.gasPriceCoef);
+
+  const body = { ...transaction.body, gas, blockRef, clauses };
```
No need to handle the case where `clauses` is still empty?

davidecarpini replied:
taking this

davidecarpini replied:
I've added an error in case of not valid recipient, but I wonder if prepareTransaction is used after the recipient is entered

![Screenshot 2023-09-12 at 10 17 09](https://github.com/LedgerHQ/ledger-live/assets/5785838/0eb388a8-be46-4ee2-9701-63e60d6a6ce3)

davidecarpini replied:
I don't think this need validation when no closures as it is like a computational function

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/js-transaction.ts:
```diff
@@ -0,0 +1,135 @@
+  account: Account,
+  transaction: Transaction,
+): Promise<Transaction> => {
+  const { amount, isTokenAccount } = await calculateTransactionInfo(account, transaction);
```
Seems like this too could be moved in the `if` condition below?

davidecarpini replied:
no because we need to return the amount below

![Screenshot 2023-09-11 at 01 41 24](https://github.com/LedgerHQ/ledger-live/assets/5785838/1c50d56c-bb3d-4ebf-a55d-c002ca48a5b2)

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/specs.ts:
```diff
@@ -0,0 +1,165 @@
+        )
+          throw new Error("no VTHO account");
+        const transaction = bridge.createTransaction(account);
+        transaction.useAllAmount = true;
```
This should be part of the `updates`

davidecarpini replied:
removed thanks

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/specs.ts:
```diff
@@ -0,0 +1,165 @@
+          throw new Error("no VTHO account");
+        const transaction = bridge.createTransaction(account);
+        transaction.useAllAmount = true;
+        transaction.subAccountId = account.subAccounts[0].id;
```
This isn't required as you already provide it as part of the `updates`, below.

davidecarpini replied:
removed thanks

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/specs.ts:
```diff
@@ -0,0 +1,165 @@
+        const transaction = bridge.createTransaction(account);
+        transaction.useAllAmount = true;
+        transaction.subAccountId = account.subAccounts[0].id;
+        const amount = account.subAccounts[0].spendableBalance.integerValue();
+        const updates = [{ amount }, { recipient }, { subAccountId: account.subAccounts[0].id }];
```
In theory you don't need to provide `amount` at all, when `useAllAmount` is used

davidecarpini replied:
I'll remove it for now

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/specs.ts:
You can apply the last comments above to all mutations in the file

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/bridge.integration.test.ts:
```diff
@@ -0,0 +1,212 @@
+          FIXME_tests: [
+            "balance is sum of ops", //the balance depends on VTHO and it's earned without operations
+            "empty transaction is equals to itself", //nonce is not deterministic
+            "ref stability on self transaction", //blockref is not deterministic
+            "can be run in parallel and all yield same results", //blockref is not deterministic
```
For both, you can now use `IgnorePrepareTransactionFields` instead

davidecarpini replied:
sorry, what do you mean by that exactly?

davidecarpini replied:
btw I tried with uncommenting them and they seem to work without adding the IgnorePrepareTransactionFields, so probably it was an old issue

davidecarpini replied:
no sorry we have the problem on ubuntu

davidecarpini replied:
I'll add them back

haammar-ledger replied:
My proposal was to use [this](https://github.com/LedgerHQ/ledger-live/blob/173d7d6d224bcf1cecf364062b6571f52792e371/libs/ledgerjs/packages/types-live/src/bridge.ts#L232).
You can find an example [here](https://github.com/LedgerHQ/ledger-live/blob/173d7d6d224bcf1cecf364062b6571f52792e371/libs/coin-evm/src/__tests__/integration/bridge.integration.test.ts#L16).

You can use it for both `nonce` and `blockref`. Doing so will just ignore those fields but still run the tests, it's less restrictive than disabling the tests.

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/account.ts:
```diff
@@ -0,0 +1,46 @@
+  return str;
+}
+
+function formatOperationSpecifics(op: Operation, unit: Unit): string {
+  const { additionalField } = op.extra as { additionalField: BigNumber };
+
+  let str = " ";
+
+  const formatConfig = {
+    disableRounding: true,
+    alwaysShowSign: false,
+    showCode: true,
+  };
+
+  str +=
+    additionalField && !additionalField.isNaN()
+      ? `\n    additionalField: ${
+          unit ? formatCurrencyUnit(unit, additionalField, formatConfig) : additionalField
+        }`
+      : "";
+
+  return str;
+}
```
This isn't needed since you don't need an `operation.extra` field

davidecarpini replied:
taking this

davidecarpini replied:
removed thanks

haammar-ledger commented on apps/ledger-live-mobile/src/components/Graph/Placeholder.tsx:
```diff
@@ -0,0 +1,36 @@
+import { Image } from "react-native";
+import { useTranslation } from "react-i18next";
+
+const graphPlaceholderSource = require("../../../assets/images/graph/graphPlaceholder.png"); // eslint-disable-line @typescript-eslint/no-var-requires
```
@KVNLS Can you please advise :pray: is this the best way to import and use a static asset? Thanks

bitkris-dev replied:
Changing it to `import` statement, but no paths aliases are provided in tsconfig for ledger-live-mobile

davidecarpini replied:
done, thanks

haammar-ledger commented on apps/ledger-live-desktop/src/renderer/screens/asset/BalanceSummary.tsx:
```diff
@@ -90,23 +95,36 @@ export default function BalanceSummary({
-          }
-          renderTooltip={renderTooltip}
-        />
+        {currency.type === "TokenCurrency" && currency.id === "vechain/vtho" ? (
+          <>
+            <Alert type="secondary" noIcon={false}>
+              <span>{t("vechain.noGraphWarning")}</span>
+            </Alert>
+            <PlaceholderChart
+              magnitude={counterValue.units[0].magnitude}
+              data={portfolio.balanceHistory}
+              tickXScale={range}
+            />
+          </>
+        ) : (
```
Can you please make this generic, in the same way you did for LLM?

We don't want any currency or token specific code outside of the `families` folder. What you did for mobile with a `tokensWithUnsupportedGraph` is acceptable though.

davidecarpini replied:
fixed thanks

haammar-ledger commented on apps/ledger-live-desktop/src/renderer/screens/account/BalanceSummary.tsx:
```diff
@@ -105,23 +111,36 @@ export default function AccountBalanceSummary({
-          }
-          renderTooltip={renderTooltip}
-        />
+        {account.type === "TokenAccount" && account.token.id === "vechain/vtho" ? (
+          <>
+            <Alert type="secondary" noIcon={false}>
+              <span>{t("vechain.noGraphWarning")}</span>
+            </Alert>
+            <PlaceholderChart
+              magnitude={counterValue.units[0].magnitude}
+              data={portfolio.balanceHistory}
+              tickXScale={range}
+            />
+          </>
+        ) : (
```
same here

davidecarpini replied:
fixed thanks

haammar-ledger reviewed (commented):
@davidecarpini Review complete :+1: 
Don't mind the number of comments, this is very clean and solid overall, thanks!

cc @hedi-edelbloute FYI

davidecarpini commented on .changeset/red-bags-own.md:
thanks removed

davidecarpini commented on apps/nanos/1.0.3/Vechain/app_1.0.7.elf:
thanks removed

davidecarpini commented on libs/ledgerjs/packages/cryptoassets/src/tokens.ts:
```diff
@@ -355,6 +358,32 @@ function convertTRONTokens(type: "trc10" | "trc20") {
 }
 
+function convertVechainToken([
+  abbr,
```
thanks changed

davidecarpini commented on libs/ledger-live-common/src/hw/signMessage/index.ts:
```diff
@@ -41,6 +41,7 @@ const signMessage: SignMessage = (transport, account, opts) => {
         `signMessage ${currency.id} on ${account.freshAddressPath} with message [${opts.message}]`,
         result,
       );
+      console.warn(result);
```
thanks removed

davidecarpini commented on libs/ledger-live-common/package.json:
```diff
@@ -259,6 +260,7 @@
     "source-map-support": "^0.5.21",
     "stellar-sdk": "^10.1.1",
     "superstruct": "0.14.2",
+    "thor-devkit": "^2.0.6",
```
Yeah, we did test it on both android and ios, we also use it on our mobile wallets. If you need to be more sure we can provide you with updated mobile builds lmk

davidecarpini commented on libs/ledger-live-common/src/families/vechain/types.ts:
```diff
@@ -0,0 +1,39 @@
+import BigNumber from "bignumber.js";
+import { Transaction as ThorTransaction } from "thor-devkit";
+
+export type NetworkInfo = {
+  family: "ethereum";
+};
```
thanks, removed

davidecarpini commented on libs/ledger-live-common/src/families/vechain/types.ts:
```diff
@@ -0,0 +1,39 @@
+export type TransactionRaw = TransactionCommonRaw & {
+  family: "vechain";
+  estimatedFees: BigNumber;
+  body: ThorTransaction.Body;
```
Confirmed, this is the structure

<img width="341" alt="Screenshot 2023-09-11 at 00 58 07" src="https://github.com/LedgerHQ/ledger-live/assets/5785838/45be5b42-bc2a-4b0c-82f7-d72ec9b7b342">

<img width="409" alt="Screenshot 2023-09-11 at 00 57 37" src="https://github.com/LedgerHQ/ledger-live/assets/5785838/7b8d37c0-83dd-45b5-84ad-cb125005b41b">

davidecarpini commented on libs/ledger-live-common/src/families/vechain/utils/calculateTransactionInfo.ts:
moved thanks!

davidecarpini commented on libs/ledger-live-common/src/families/vechain/utils/transaction-utils.ts:
```diff
@@ -0,0 +1,104 @@
+  return new BigNumber(baseGasPrice).times(gasPriceCoef).idiv(255).plus(baseGasPrice).times(gas);
+};
+
+/**
+ * Get fees paid for the transaction
+ * @param transactionId - the id of the transaction
+ * @return the fee paid in VTHO or 0
+ */
+export const getFees = async (transactionID: string): Promise<BigNumber> => {
+  const { data } = await network({
+    method: "GET",
+    url: `${BASE_URL}/transactions/${transactionID}/receipt`,
+    params: { id: transactionID },
+  });
+
+  if (!data || !data.paid) return new BigNumber(0);
+  return new BigNumber(data.paid);
+};
```
moved thanks

davidecarpini commented on libs/ledger-live-common/src/families/vechain/js-transaction.ts:
```diff
@@ -0,0 +1,135 @@
+  return { ...transaction, body, amount, estimatedFees };
+};
+
+export const calculateClausesVtho = async (
```
Unfortunately, it needs it because it returns an array of promises and ts raises an error:

![Screenshot 2023-09-11 at 01 12 53](https://github.com/LedgerHQ/ledger-live/assets/5785838/ca4d6687-d06e-4f67-97fb-d5f9702ac9fb)

davidecarpini commented on libs/ledger-live-common/src/families/vechain/js-transaction.ts:
```diff
@@ -0,0 +1,135 @@
+  return clauses;
+};
+
+const calculateClausesVet = async (
```
same

davidecarpini commented on libs/ledger-live-common/src/families/vechain/utils/calculateTransactionInfo.ts:
```diff
@@ -0,0 +1,88 @@
+  //   const estimatedFees = new BigNumber(gas);
+  //   return estimatedFees;
+  // }
+  const fees = calculateFee(new BigNumber("67000"), DEFAULT_GAS_COEFFICIENT);
```
it doesn't make much difference because it still returns a promise, but I've added it

davidecarpini commented on libs/ledger-live-common/src/families/vechain/specs.ts:
```diff
@@ -0,0 +1,165 @@
+  appQuery: {
+    model: DeviceModelId.nanoS,
+    appName: "Vechain",
+    appVersion: "1.0.7",
```
tbh I don't know why mirko added the version, removed it for now 👍

davidecarpini commented on libs/ledger-live-common/src/families/vechain/bridge/js.ts:
```diff
@@ -0,0 +1,54 @@
+
+const currencyBridge: CurrencyBridge = {
+  scanAccounts,
+  preload: async (): Promise<Record<string, any>> => {
+    return {};
+  },
```
thanks changed

davidecarpini commented on libs/ledger-live-common/src/families/vechain/bridge/js.ts:
```diff
@@ -0,0 +1,54 @@
+  hydrate: (): void => {},
+};
+
+const estimateMaxSpendable = async (inputs: {
+  account: AccountLike;
+  transaction: Transaction;
+}): Promise<BigNumber> => {
+  const { account, transaction } = inputs;
+
+  if (account.type === "Account") {
+    return account.balance;
+  }
+  if (transaction) {
+    const maxTokenFees = await calculateMaxFeesToken();
+    const spendable = account.balance.minus(maxTokenFees);
+    if (spendable.gt(0)) return account.balance.minus(maxTokenFees);
+    return new BigNumber(0);
+  } else {
+    return account.balance;
+  }
+};
```
moved thanks!

davidecarpini commented on libs/ledger-live-common/src/families/vechain/js-getTransactionStatus.ts:
```diff
@@ -0,0 +1,71 @@
+  return Promise.resolve({
+    errors,
+    warnings,
+    estimatedFees: Object.keys(errors).length ? new BigNumber(0) : transaction.estimatedFees,
+    amount: amount,
+    totalSpent: Object.keys(errors).length ? new BigNumber(0) : totalSpent,
```
if I remember correctly it's because it could return the wrong amount if there are some errors

davidecarpini replied:
removed we don't need them, thanks!

davidecarpini commented on libs/ledger-live-common/src/families/vechain/js-signOperation.ts:
```diff
@@ -0,0 +1,103 @@
+            account.freshAddressPath,
+            unsigned.encode().toString("hex"),
+          );
+          o.next({
+            type: "device-signature-requested",
+          });
```
moved thanks

davidecarpini commented on libs/ledger-live-common/src/families/vechain/types.ts:
```diff
@@ -0,0 +1,39 @@
+  family: "vechain";
+  estimatedFees: BigNumber;
+  body: ThorTransaction.Body;
+  // networkInfo property is just a work-around to display fees on mobile, more info here: apps/ledger-live-mobile/src/components/SendRowsFee.tsx
+  networkInfo: boolean;
```
taking this

davidecarpini replied:
thanks removed

davidecarpini commented on libs/ledger-live-common/src/families/vechain/transaction.ts:
```diff
@@ -0,0 +1,58 @@
+  toTransactionStatusRawCommon,
+} from "@ledgerhq/coin-framework/transaction/common";
+
+export const formatTransaction = (t: Transaction): string => {
+  return t.estimatedFees.toString();
+};
```
taking this

davidecarpini replied:
added taking as example stacks coin that seems to be a recent one, thanks

davidecarpini commented on libs/ledger-live-common/src/families/vechain/types.ts:
```diff
@@ -0,0 +1,39 @@
+
+export type TransactionRaw = TransactionCommonRaw & {
+  family: "vechain";
+  estimatedFees: BigNumber;
```
taking this one

davidecarpini replied:
fixed thanks

davidecarpini commented on libs/ledger-live-common/src/families/vechain/transaction.ts:
```diff
@@ -0,0 +1,58 @@
+  return t.estimatedFees.toString();
+};
+
+export const fromTransactionRaw = (tr: TransactionRaw): Transaction => {
+  const common = fromTransactionCommonRaw(tr);
+
+  return {
+    networkInfo: true,
+    ...tr,
+    ...common,
+  };
+};
+
+export const toTransactionRaw = (t: Transaction): TransactionRaw => {
+  const common = toTransactionCommonRaw(t);
+
+  return {
+    ...t,
+    ...common,
+  };
+};
```
I've changed estimatedFees to string, and any other fields are serializable

![Screenshot 2023-09-11 at 11 42 02](https://github.com/LedgerHQ/ledger-live/assets/5785838/df455fcb-6df4-4a33-a5c5-f12eccc9adf3)

![Screenshot 2023-09-11 at 11 42 08](https://github.com/LedgerHQ/ledger-live/assets/5785838/0db41d37-1b02-42d0-923e-a72e9c39e2e1)

davidecarpini commented on libs/ledger-live-common/src/families/vechain/js-transaction.ts:
```diff
@@ -0,0 +1,135 @@
+ * @param {TransactionCommon} patch
+ * @returns patched transaction
+ */
+export const updateTransaction = (
+  t: Transaction,
+  patch: $Shape<TransactionCommon>,
+): Transaction => {
+  return { ...t, ...patch };
+};
```
thanks removed

davidecarpini commented on libs/ledger-live-common/src/families/vechain/api/sdk.ts:
```diff
@@ -0,0 +1,144 @@
+    order: "desc",
+  };
+
+  const { data } = await network({
```
taking this

davidecarpini replied:
get feedback, yeah it will return every operation

davidecarpini commented on libs/ledger-live-common/src/families/vechain/cli-transaction.ts:
```diff
@@ -0,0 +1,66 @@
+
+    const clauses: Array<Clauses> = [];
+
+    if (opts.vechainCurrency == "VET") {
+      clauses.push({
+        to: transaction.recipient,
+        value: "0x" + transaction.amount.toString(16),
+        data: "0x",
+      });
+    } else if (opts.vechainCurrency == "VTHO") {
+      clauses.push({
+        value: 0,
+        to: VTHO_ADDRESS,
+        data: VIP180.transfer.encode(transaction.recipient, transaction.amount.toFixed()),
+      });
+    }
```
taking this

davidecarpini commented on libs/ledger-live-common/src/families/vechain/js-signOperation.ts:
```diff
@@ -0,0 +1,103 @@
+            signedOperation: {
+              operation,
+              signature: signature.toString("hex"),
+              expirationDate: undefined,
```
removed thanks

davidecarpini commented on libs/ledger-live-common/src/families/vechain/js-signOperation.ts:
```diff
@@ -0,0 +1,103 @@
+    recipients: [transaction.recipient].filter(Boolean),
+    accountId: account.id,
+    date: new Date(),
+    extra: { transaction },
```
changed thanks!

davidecarpini commented on libs/ledger-live-common/src/families/vechain/bridge.integration.test.ts:
```diff
@@ -0,0 +1,212 @@
+  implementations: ["js"],
+  currencies: {
+    vechain: {
+      FIXME_ignoreAccountFields: ["balance", "spendableBalance", "estimateMaxSpendable"],
```
tbh I don't know, mirko did that, I'll try to remove them and see if it changes anything

davidecarpini replied:
adding them back

davidecarpini replied:
this is because VTHO balance increase when you have VET on your account 
more on this:
https://docs.vechain.org/thor/learn/two-token-design.html

davidecarpini commented on libs/ledger-live-common/src/families/vechain/js-signOperation.ts:
```diff
@@ -0,0 +1,103 @@
+    subOperations: subAccountId
+      ? [
+          {
+            id: `${subAccountId}--OUT`,
```
taking this

davidecarpini replied:
changed thanks

davidecarpini commented on libs/ledger-live-common/src/families/vechain/bridge.integration.test.ts:
```diff
@@ -0,0 +1,212 @@
+        {
+          raw: vechain1 as AccountRaw,
+          implementations: ["js"],
+          transactions: [
```
I do not have strong knowledge of bridge tests but the NotEnoughVTHO depends on the balance of the account, 
Since the average fees are about 0.21 vtho, it would mean that we need to have an account with some vet and 0 vtho, can we reach that, with an isolated test, replicating this every time?

davidecarpini replied:
added NotEnoughVTHO case 👍

hedi-edelbloute commented on apps/ledger-live-desktop/src/renderer/screens/account/BalanceSummary.tsx:
```diff
@@ -105,23 +111,36 @@ export default function AccountBalanceSummary({
-          }
-          renderTooltip={renderTooltip}
-        />
+        {account.type === "TokenAccount" && account.token.id === "vechain/vtho" ? (
```
We do not want vechain specific code here, sorry, please check this for guidelines : https://developers.ledger.com/docs/coin/ui-ux-guidelines/

davidecarpini replied:
yeah the fix arriving, @haammar-ledger asked that too

hedi-edelbloute commented on apps/ledger-live-desktop/src/renderer/screens/asset/BalanceSummary.tsx:
```diff
@@ -90,23 +95,36 @@ export default function BalanceSummary({
-          }
-          renderTooltip={renderTooltip}
-        />
+        {currency.type === "TokenCurrency" && currency.id === "vechain/vtho" ? (
```
same as account/BalanceSummary : https://developers.ledger.com/docs/coin/ui-ux-guidelines/

davidecarpini replied:
same

davidecarpini commented on libs/ledger-live-common/src/families/vechain/bridge/js.ts:
thanks, added

davidecarpini commented on libs/ledger-live-common/src/families/vechain/bridge.integration.test.ts:
```diff
@@ -0,0 +1,212 @@
+      accounts: [
+        {
+          raw: vechain1 as AccountRaw,
+          implementations: ["js"],
```
tested and it worked, thanks!

hedi-edelbloute commented on libs/ledger-live-common/src/families/vechain/js-synchronisation.ts:
```diff
@@ -0,0 +1,90 @@
+  const vthoOperations = await getTokenOperations(
+    vthoAccountId,
+    address,
+    "0x0000000000000000000000000000456e65726779",
```
I understand this is VTHO token address, maybe we could store it in a const variable named something like VTHO_TOKEN_ADDRESS

minor

davidecarpini replied:
done thanks

hedi-edelbloute commented on libs/ledger-live-common/src/families/vechain/js-synchronisation.ts:
```diff
@@ -0,0 +1,90 @@
+    vthoAccountId,
+    address,
+    "0x0000000000000000000000000000456e65726779",
+    1,
```
Maybe it would make sense to start at account's first operation block ? Ignore if it could cause problems

bitkris-dev replied:
Sorry I don't get what you mean

davidecarpini replied:
I'll change it since  the first block seems to be 0, don't know why we set 1

davidecarpini replied:
<img width="901" alt="Screenshot 2023-09-25 at 11 03 25" src="https://github.com/LedgerHQ/ledger-live/assets/5785838/782d2917-46fb-4d98-8859-546f1790232f">

davidecarpini replied:
https://docs.vechain.org/thor/thorest-api

hedi-edelbloute commented on libs/ledger-live-common/src/families/vechain/utils/transaction-utils.ts:
```diff
@@ -0,0 +1,155 @@
+};
+
+export const calculateMaxFeesToken = async (): Promise<BigNumber> => {
+  //FIXME: The BE call is currently returning not stable values, hardwiring a value
```
Will it ever be fixed ? I think we need to choose between dynamic value and static value as it is implemented here

davidecarpini replied:
asking it, thanks

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/utils/mapping-utils.ts:
```diff
@@ -0,0 +1,93 @@
+ * @param roundingStrategy - what strategy to use when rounding. Based on the strategies defined in `bignumber.js`. Default strategy is ROUND_HALF_UP
+ * @returns the scaled up result as a string
+ */
+export const scaleNumberUp = (
```
This function doesn't seem to be used anywhere, I think it can be removed.

davidecarpini replied:
thanks I'll remove it

davidecarpini replied:
done

haammar-ledger commented on libs/ledger-live-common/src/families/vechain/utils/transaction-utils.ts:
```diff
@@ -0,0 +1,230 @@
+  return new BigNumber(baseGasPrice).times(gasPriceCoef).idiv(255).plus(baseGasPrice).times(gas);
+};
+
+export const calculateTransactionEconomicInfo = async ({
```
Same for this one, is it used anywhere?

davidecarpini replied:
thanks I'll remove it

davidecarpini replied:
done

haammar-ledger reviewed (approved):
LGTM, only minor suggestions if you can have a look

davidecarpini commented on libs/ledger-live-common/src/families/vechain/specs.ts:
```diff
@@ -0,0 +1,171 @@
+  name: "Vechain",
+  currency: getCryptoCurrencyById("vechain"),
+  appQuery: {
+    model: DeviceModelId.nanoS,
```
I tried but I got this error:
```
debug: hw: Initial error remapping {"error":{"message":"error request aborted","name":"Error","stack":"Error: error request aborted\n    at createError (/Users/macbook/ledger-live/node_modules/.pnpm/axios@0.26.1/node_modules/axios/lib/core/createError.js:16:15)\n    at IncomingMessage.handlerStreamAborted (/Users/macbook/ledger-live/node_modules/.pnpm/axios@0.26.1/node_modules/axios/lib/adapters/http.js:301:18)\n    at IncomingMessage.emit (node:events:527:28)\n    at IncomingMessage._destroy (node:_http_incoming:179:10)\n    at _destroy (node:internal/streams/destroy:102:25)\n    at IncomingMessage.destroy (node:internal/streams/destroy:64:5)\n    at Socket.socketCloseListener (node:_http_client:414:11)\n    at Socket.emit (node:events:539:35)\n    at TCP.<anonymous> (node:net:709:12)","config":{"transitional":{"silentJSONParsing":true,"forcedJSONParsing":true,"clarifyTimeoutError":false},"transformRequest":[null],"transformResponse":[null],"timeout":0,"xsrfCookieName":"XSRF-TOKEN","xsrfHeaderName":"X-XSRF-TOKEN","maxContentLength":-1,"maxBodyLength":-1,"headers":{"Accept":"application/json, text/plain, */*","X-Ledger-Client-Version":"cli/0.0.0","Content-Type":"application/json","User-Agent":"axios/0.26.1","Content-Length":63},"httpsAgent":{"_events":{},"_eventsCount":2,"defaultPort":443,"protocol":"https:","options":{"keepAlive":true,"path":null},"requests":{},"sockets":{},"freeSockets":{},"keepAliveMsecs":1000,"keepAlive":true,"maxSockets":null,"maxFreeSockets":256,"scheduling":"lifo","maxTotalSockets":null,"totalSocketCount":0,"maxCachedSessions":100,"_sessionCache":{"map":{},"list":[]}},"baseURL":"http://localhost:30001","method":"post","url":"/apdu","data":"{\"data\":\"e002000015058000002c80000332800000000000000000000000\"}"},"code":"ERR_REQUEST_ABORTED","status":null},"mappedError":{"message":"error request aborted","name":"Error","stack":"Error: error request aborted\n    at createError (/Users/macbook/ledger-live/node_modules/.pnpm/axios@0.26.1/node_modules/axios/lib/core/createError.js:16:15)\n    at IncomingMessage.handlerStreamAborted (/Users/macbook/ledger-live/node_modules/.pnpm/axios@0.26.1/node_modules/axios/lib/adapters/http.js:301:18)\n    at IncomingMessage.emit (node:events:527:28)\n    at IncomingMessage._destroy (node:_http_incoming:179:10)\n    at _destroy (node:internal/streams/destroy:102:25)\n    at IncomingMessage.destroy (node:internal/streams/destroy:64:5)\n    at Socket.socketCloseListener (node:_http_client:414:11)\n    at Socket.emit (node:events:539:35)\n    at TCP.<anonymous> (node:net:709:12)","config":{"transitional":{"silentJSONParsing":true,"forcedJSONParsing":true,"clarifyTimeoutError":false},"transformRequest":[null],"transformResponse":[null],"timeout":0,"xsrfCookieName":"XSRF-TOKEN","xsrfHeaderName":"X-XSRF-TOKEN","maxContentLength":-1,"maxBodyLength":-1,"headers":{"Accept":"application/json, text/plain, */*","X-Ledger-Client-Version":"cli/0.0.0","Content-Type":"application/json","User-Agent":"axios/0.26.1","Content-Length":63},"httpsAgent":{"_events":{},"_eventsCount":2,"defaultPort":443,"protocol":"https:","options":{"keepAlive":true,"path":null},"requests":{},"sockets":{},"freeSockets":{},"keepAliveMsecs":1000,"keepAlive":true,"maxSockets":null,"maxFreeSockets":256,"scheduling":"lifo","maxTotalSockets":null,"totalSocketCount":0,"maxCachedSessions":100,"_sessionCache":{"map":{},"list":[]}},"baseURL":"http://localhost:30001","method":"post","url":"/apdu","data":"{\"data\":\"e002000015058000002c80000332800000000000000000000000\"}"},"code":"ERR_REQUEST_ABORTED","status":null}}
debug: hw: Closing and cleaning transport
debug: hw: Trying to close transport
debug: hw: Closing transport speculos
debug: hw: Job error {"error":{"message":"error request aborted","name":"Error","stack":"Error: error request aborted\n    at createError (/Users/macbook/ledger-live/node_modules/.pnpm/axios@0.26.1/node_modules/axios/lib/core/createError.js:16:15)\n    at IncomingMessage.handlerStreamAborted (/Users/macbook/ledger-live/node_modules/.pnpm/axios@0.26.1/node_modules/axios/lib/adapters/http.js:301:18)\n    at IncomingMessage.emit (node:events:527:28)\n    at IncomingMessage._destroy (node:_http_incoming:179:10)\n    at _destroy (node:internal/streams/destroy:102:25)\n    at IncomingMessage.destroy (node:internal/streams/destroy:64:5)\n    at Socket.socketCloseListener (node:_http_client:414:11)\n    at Socket.emit (node:events:539:35)\n    at TCP.<anonymous> (node:net:709:12)","config":{"transitional":{"silentJSONParsing":true,"forcedJSONParsing":true,"clarifyTimeoutError":false},"transformRequest":[null],"transformResponse":[null],"timeout":0,"xsrfCookieName":"XSRF-TOKEN","xsrfHeaderName":"X-XSRF-TOKEN","maxContentLength":-1,"maxBodyLength":-1,"headers":{"Accept":"application/json, text/plain, */*","X-Ledger-Client-Version":"cli/0.0.0","Content-Type":"application/json","User-Agent":"axios/0.26.1","Content-Length":63},"httpsAgent":{"_events":{},"_eventsCount":2,"defaultPort":443,"protocol":"https:","options":{"keepAlive":true,"path":null},"requests":{},"sockets":{},"freeSockets":{},"keepAliveMsecs":1000,"keepAlive":true,"maxSockets":null,"maxFreeSockets":256,"scheduling":"lifo","maxTotalSockets":null,"totalSocketCount":0,"maxCachedSessions":100,"_sessionCache":{"map":{},"list":[]}},"baseURL":"http://localhost:30001","method":"post","url":"/apdu","data":"{\"data\":\"e002000015058000002c80000332800000000000000000000000\"}"},"code":"ERR_REQUEST_ABORTED","status":null}}
```
leaving nanoS for now

davidecarpini commented on libs/ledger-live-common/src/families/vechain/bridge/js-estimateMaxSpendable.ts:
```diff
@@ -0,0 +1,23 @@
+    return new BigNumber(0);
+  } else {
+    return account.balance;
+  }
```
done

davidecarpini commented:
> Did you forget to push the changes maybe? It doesn't seem that the comments in this file have been addressed?
@haammar-ledger  I added all of them, but then I reverted the ones that were breaking the bot test
